tools/functional-tester/etcd-tester/lease_stresser.go

```diff
@@ -0,0 +1,281 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/etcdserver/api/v3rpc/rpctypes"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+// time to live for lease
+const TTL = 15
+
+type leaseStressConfig struct {
+	numOfLeases  int
+	keysPerLease int
+}
+
+type leaseStresser struct {
+	endpoint string
+	cancel   func()
+	conn     *grpc.ClientConn
+	kvc      pb.KVClient
+	lc       pb.LeaseClient
+
+	success      int
+	failure      int
+	numOfLeases  int
+	keysPerLease int
+
+	// mu Mutex is used to protect the shared maps aliveLeases and revokedLeases
+	// which are accessed and modified by different go routines.
+	mu            sync.Mutex
+	aliveLeases   map[int64]struct{}
+	revokedLeases map[int64]struct{}
+
+	runWg   *sync.WaitGroup
+	aliveWg *sync.WaitGroup
+}
+
+type leaseStresserBuilder func(m *member) Stresser
+
+func newLeaseStressBuilder(lsConfig *leaseStressConfig) leaseStresserBuilder {
+	return func(mem *member) Stresser {
+		return &leaseStresser{
+			endpoint:     mem.grpcAddr(),
+			numOfLeases:  lsConfig.numOfLeases,
+			keysPerLease: lsConfig.keysPerLease,
+		}
+	}
+}
+
+func (ls *leaseStresser) Setup() error {
+	if ls.numOfLeases == 0 {
+		panic("expect numOfLeases to be set")
+	}
+	if ls.keysPerLease == 0 {
+		panic("expect keysPerLease to be set")
+	}
+	conn, err := grpc.Dial(ls.endpoint, grpc.WithInsecure())
+	if err != nil {
+		return fmt.Errorf("%v (%s)", err, ls.endpoint)
+	}
+	ls.runWg = &sync.WaitGroup{}
+	ls.aliveWg = &sync.WaitGroup{}
+	ls.kvc = pb.NewKVClient(conn)
+	ls.lc = pb.NewLeaseClient(conn)
+	ls.conn = conn
+	ls.aliveLeases = make(map[int64]struct{})
+	ls.revokedLeases = make(map[int64]struct{})
+	return nil
+}
+
+func (ls *leaseStresser) Stress() error {
+	plog.Infof("lease Stresser %v starting ...", ls.endpoint)
+	ctx, cancel := context.WithCancel(context.Background())
+	ls.cancel = cancel
+	ls.runWg.Add(1)
+	go ls.run(ctx)
+	return nil
+}
+
+func (ls *leaseStresser) run(ctx context.Context) {
+	defer ls.runWg.Done()
+	ls.createLeases(ctx)
+	ls.randomlyDropLeases(ctx)
+}
+
+func (ls *leaseStresser) createLeases(ctx context.Context) {
+	numOfLeasesToCreate := ls.numOfLeases - len(ls.aliveLeases)
+	var wg sync.WaitGroup
+	wg.Add(numOfLeasesToCreate)
+	for i := 0; i < numOfLeasesToCreate; i++ {
+		go func() {
+			defer wg.Done()
+			leaseID, err := ls.createLeaseAndKeepAlive(ctx)
+			if err != nil {
+				plog.Errorf("Lease creation error %v", err)
+				return
+			}
+			plog.Debugf("Lease %v created ", leaseID)
+			ls.attachKeysWithLease(ctx, leaseID)
```

Comment:
it can't fail because i keep sending Put request until I receive a certainty from the server.

```go
func (ls *leaseStresser) attachKeysWithLease(ctx context.Context, leaseID int64) {
	for j := 0; j < ls.keysPerLease; j++ {
		// keep retrying until put succeeded for sure assuming etcd servers will be recovered eventually
		for {
			_, err := ls.kvc.Put(ctx, &pb.PutRequest{
				Key:   []byte(fmt.Sprintf("%d%s%d", leaseID, "_", j)),
				Value: []byte(fmt.Sprintf("bar")),
				Lease: leaseID,
			})
			if err == nil {
				return
			}
		}
	}
}
```